### PR TITLE
Closes #54507 - Adds themable window border

### LINKED ISSUE #54507: Add support for themable window border
On Windows 10 at least with the new custom titlebar (`"window.titleBarStyle": "custom"`) vscode loses its window border. It would be great for a theme to be able to specify a border color in this case (if desired). Because with some themes (particularly mine) it can be hard to tell where 1 window ends and another one starts.

I tried setting `titleBar.border` but it wasn't what I was looking for:
image

//cc @sbatten

### PR DESCRIPTION
**Pull request by eamodio**
Closes #54507 

@bpasero Here is the first pass at adding a themable window border. My open questions are:

- Is the split between `workbench.ts` and `layout.ts` correct? Or should I move more into `layout.ts`
- Is it OK to switch to use `vh/vw` units for sizing, rather than the manual pixel sizing we currently have?

Thoughts?

### CODE REVIEW

File: src/vs/workbench/browser/workbench.ts [hunk 1]
```diff
@@ -462,4 +524,9 @@ export class Workbench extends Layout {
 			mark('didStartWorkbench');
 		}
 	}
+
+	getClientArea(): Dimension {
+		const dim = super.getClientArea();
+		return this.windowBorder ? new Dimension(dim.width - 2, dim.height - 2) : dim;
```
Comment (eamodio) [inline src/vs/workbench/browser/workbench.ts]: Can't say I love this, but...

**Comment by eamodio** [comment]
You can see the border in action by adding the following to your user `settings.json`

```
    "workbench.colorCustomizations": {
        "window.activeBorder": "#00ff00",
        "window.inactiveBorder": "#ff0000"
    }
```

**Comment by AlexFiliakov** [comment]
This would be super useful for me, as I have a 4k monitor and often spawn multiple VSCode sessions to move between related projects. The windows often blend together, which is a nuisance.

For example, here is one VSCode window on top of another, but it's hard to tell where one ends and the other begins:
Overlay VSCode Screenshot

**Comment by eamodio** [comment]
@bpasero I moved the logic into `layout.ts` -- its not clear to me what could/should be moved into `window.ts`. I also added a first pass at a border radius on a mac, so hopefully that will help (I don't have a mac yet to test it on)

File: src/vs/workbench/common/theme.ts [hunk 2]
```diff
@@ -589,6 +589,18 @@ export const NOTIFICATIONS_INFO_ICON_FOREGROUND = registerColor('notificationsIn
 	hc: editorInfoForeground
 }, nls.localize('notificationsInfoIconForeground', "The color used for the notification info icon."));
 
+export const WINDOW_ACTIVE_BORDER = registerColor('window.activeBorder', {
```
Comment (bpasero) [inline src/vs/workbench/common/theme.ts]: @eamodio unfortunately this means the color will appear for web users in their settings, so at least maybe add a documentation that this is not supported in web?

File: src/vs/workbench/browser/media/style.css [hunk 3]
```diff
@@ -51,6 +51,23 @@ body.web {
 	position: relative;
 	z-index: 1;
 	overflow: hidden;
+	height: 100vh;
```
Comment (bpasero) [inline src/vs/workbench/browser/media/style.css]: @eamodio this change is not clear to me and I am not sure about the consequences. Is it really needed and why this format and not `100%`?

File: src/vs/workbench/browser/media/style.css [hunk 4]
```diff
@@ -51,6 +51,23 @@ body.web {
...
+	height: calc(100vh - 2px);
+	width: calc(100vw - 2px);
+	border: 1px solid var(--window-border-color);
+}
+
+.monaco-workbench.border.mac {
+	border-radius: 5px;
+}
+
+.monaco-workbench.border.windows {
```
Comment (bpasero) [inline src/vs/workbench/browser/media/style.css]: @eamodio why is this rule needed?

File: src/vs/workbench/browser/layout.ts [hunk 5]
```diff
@@ -257,6 +264,12 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 		if ((isWindows || isLinux || isWeb) && getTitleBarStyle(this.configurationService, this.environmentService) === 'custom') {
 			this._register(this.titleService.onMenubarVisibilityChange(visible => this.onMenubarToggled(visible)));
 		}
+
+		// Theme changes
+		this._register(this.themeService.onThemeChange(() => this.setWindowBorder()));
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @eamodio I would try to align with what our `Themable` class does (which unfortunately we cannot extend here) and introduce a method `updateStyles()` which then calls `this.setWindowBorder()`.

File: src/vs/workbench/browser/layout.ts [hunk 6]
```diff
@@ -826,12 +892,14 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
...
+	getClientArea(): Dimension {
+		const dim = getClientArea(this.parent);
+		return this.state.windowBorder ? new Dimension(dim.width - 2, dim.height - 2) : dim;
+	}
+
 	layout(): void {
 		if (!this.disposed) {
-			this._dimension = getClientArea(this.parent);
-
-			position(this.container, 0, 0, 0, 0, 'relative');
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: @eamodio why is this removed?

**Review by bpasero — CHANGES_REQUESTED** [review]
@eamodio yeah this is better. I think given the layout needs to know if a border is set, it is fine to leave the changes in `layout.ts`.

File: src/vs/workbench/browser/media/style.css (hunk 3 above)
Comment (eamodio) [reply src/vs/workbench/browser/media/style.css]: By using `vh/wh` we can avoid having to manually size the container and avoids other issues with using % where all the parents need to be 100%, etc. So 100vh/wh will always size the container to the viewport height/width.

File: src/vs/workbench/browser/media/style.css (hunk 4 above)
Comment (eamodio) [reply src/vs/workbench/browser/media/style.css]: It isn't and can be removed.

File: src/vs/workbench/browser/layout.ts (hunk 5 above)
Comment (eamodio) [reply src/vs/workbench/browser/layout.ts]: Will do

File: src/vs/workbench/browser/layout.ts (hunk 6 above)
Comment (eamodio) [reply src/vs/workbench/browser/layout.ts]: By using `vh/wh` in the CSS we don't need to manually size it any more

File: src/vs/workbench/browser/media/style.css (hunk 3 above)
Comment (bpasero) [reply src/vs/workbench/browser/media/style.css]: @eamodio still, we used to set `position: relative` programmatically, no?

File: src/vs/workbench/common/theme.ts (hunk 2 above)
Comment (eamodio) [reply src/vs/workbench/common/theme.ts]: Will do

File: src/vs/workbench/browser/media/style.css (hunk 3 above)
Comment (eamodio) [reply src/vs/workbench/browser/media/style.css]: We did, but it is set on `monaco-workbench` -- is there a need for the inline-style?

File: src/vs/workbench/browser/layout.ts [hunk 7]
```diff
@@ -257,6 +264,12 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 		if ((isWindows || isLinux || isWeb) && getTitleBarStyle(this.configurationService, this.environmentService) === 'custom') {
 			this._register(this.titleService.onMenubarVisibilityChange(visible => this.onMenubarToggled(visible)));
 		}
+
+		// Theme changes
+		this._register(this.themeService.onThemeChange(theme => this.onThemeChanged(theme)));
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]:
```suggestion
		this._register(this.themeService.onThemeChange(theme => this.updateStyles()));
```

File: src/vs/workbench/browser/layout.ts [hunk 8]
```diff
@@ -297,6 +310,19 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 		this._onFullscreenChange.fire(this.state.fullscreen);
 	}
 
+	private onThemeChanged(theme: ITheme) {
+		this.updateStyles();
+	}
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: Remove

File: src/vs/workbench/browser/layout.ts [hunk 9]
```diff
@@ -366,6 +392,49 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
 		this.layout();
 	}
 
+	private setWindowBorder(skipLayout: boolean = false) {
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: Suggest: `updateWindowBorder`

File: src/vs/workbench/browser/layout.ts [hunk 10]
```diff
@@ -366,6 +392,49 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
...
+		if (windowBorder === this.state.windowBorder) {
+			return;
+		}
+
+		this.state.windowBorder = windowBorder;
+
+		if (windowBorder) {
+			addClass(this.container, Classes.WINDOW_BORDER);
+		}
+		else {
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: Our coding style is to put the `else` statement on the same line. Even better, I suggest to use `toggleClass` helper to remove the if/else.

File: src/vs/workbench/browser/layout.ts [hunk 11]
```diff
@@ -366,6 +392,49 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
...
+
+		const activeBorder = theme.getColor(WINDOW_ACTIVE_BORDER);
+		const inactiveBorder = theme.getColor(WINDOW_INACTIVE_BORDER);
+
+		let windowBorder = false;
+		if (activeBorder || inactiveBorder) {
+			windowBorder = true;
+
+			let borderColor = this.state.hasFocus ? activeBorder : inactiveBorder;
+			if (!borderColor) {
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: I do not really understand this fallback to the title color. When I tested this on mac I think I saw a white border around my window. Why do we need this?

File: src/vs/workbench/browser/layout.ts [hunk 12]
```diff
@@ -366,6 +392,49 @@ export abstract class Layout extends Disposable implements IWorkbenchLayoutServi
...
+				borderColor = theme.getColor(this.state.hasFocus ? TITLE_BAR_ACTIVE_BACKGROUND : TITLE_BAR_INACTIVE_BACKGROUND);
+			}
+			this.container.style.setProperty('--window-border-color', borderColor ? borderColor.toString() : 'transparent');
```
Comment (bpasero) [inline src/vs/workbench/browser/layout.ts]: All in all I think I would really not do any border business if no color is defined.

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/browser/media/style.css (hunk 3 above)
Comment (bpasero) [reply src/vs/workbench/browser/media/style.css]: True, ignore.

**Comment by bpasero** [comment]
As requested, looks on macOS:

image

File: src/vs/workbench/browser/layout.ts (hunk 11 above)
Comment (eamodio) [reply src/vs/workbench/browser/layout.ts]: I added that fallback, because if there is only 1 color defined (active or inactive), then when the window is focused/unfocused the border would jump on and off.

File: src/vs/workbench/browser/layout.ts (hunk 12 above)
Comment (eamodio) [reply src/vs/workbench/browser/layout.ts]: I removed the fallback to the titlebar color, and instead if one color is missing, it will just fallback to the other one.

File: src/vs/workbench/browser/layout.ts (hunk 11 above)
Comment (bpasero) [reply src/vs/workbench/browser/layout.ts]: @eamodio I understand, but in that case wouldn't it be better to use the one color that is defined and not some random other color or even transparent?

Comment (eamodio) [reply src/vs/workbench/browser/layout.ts]: Yeah, the new code uses the 1 color that is defined

**Review by bpasero — APPROVED** [review]
